crates/composefs/src/splitstream.rs

```diff
@@ -6,172 +6,258 @@
+    collections::{BTreeMap, HashMap},
+    fs::File,
+    hash::Hash,
+    io::{BufRead, BufReader, Read, Seek, SeekFrom, Take, Write},
+    mem::size_of,
+    mem::MaybeUninit,
+    ops::Range,
     sync::Arc,
 };
 
-use anyhow::{bail, Result};
-use sha2::{Digest, Sha256};
-use zerocopy::{FromBytes, Immutable, IntoBytes, KnownLayout};
+use anyhow::{bail, ensure, Context, Error, Result};
+use rustix::{
+    buffer::spare_capacity,
+    io::{pread, read},
+};
+use zerocopy::{
+    little_endian::{I64, U16, U64},
+    FromBytes, Immutable, IntoBytes, KnownLayout,
+};
 use zstd::stream::{read::Decoder, write::Encoder};
 
-use crate::{
-    fsverity::FsVerityHashValue,
-    repository::Repository,
-    util::{read_exactish, Sha256Digest},
-};
+use crate::{fsverity::FsVerityHashValue, repository::Repository, util::read_exactish};
+
+const SPLITSTREAM_MAGIC: [u8; 11] = *b"SplitStream";
+const LG_BLOCKSIZE: u8 = 12; // TODO: hard-coded 4k.  make this generic later...
 
-/// A single entry in the digest map, mapping content SHA256 hash to fs-verity object ID.
+// Nearly everything in the file is located at an offset indicated by a FileRange.
+#[derive(Debug, Clone, Copy, FromBytes, Immutable, IntoBytes, KnownLayout)]
+struct FileRange {
+    start: U64,
+    end: U64,
+}
+
+// The only exception is the header: it is a fixed sized and comes at the start (offset 0).
 #[derive(Debug, FromBytes, Immutable, IntoBytes, KnownLayout)]
-#[repr(C)]
-pub struct DigestMapEntry<ObjectID: FsVerityHashValue> {
-    /// SHA256 hash of the content body
-    pub body: Sha256Digest,
-    /// fs-verity object identifier
-    pub verity: ObjectID,
+struct SplitstreamHeader {
+    pub magic: [u8; 11],  // Contains SPLITSTREAM_MAGIC
+    pub version: u8,      // must always be 0
+    pub _flags: U16,      // is currently always 0 (but ignored)
+    pub algorithm: u8,    // kernel fs-verity algorithm identifier (1 = sha256, 2 = sha512)
+    pub lg_blocksize: u8, // log2 of the fs-verity block size (12 = 4k, 16 = 64k)
+    pub info: FileRange,  // can be used to expand/move the info section in the future
 }
 
-/// A map of content digests to object IDs, maintained in sorted order for binary search.
-#[derive(Debug)]
-pub struct DigestMap<ObjectID: FsVerityHashValue> {
-    /// Vector of digest map entries, kept sorted by body hash
-    pub map: Vec<DigestMapEntry<ObjectID>>,
+// The info block can be located anywhere, indicated by the "info" FileRange in the header.
+#[derive(Debug, FromBytes, Immutable, IntoBytes, KnownLayout)]
+struct SplitstreamInfo {
+    pub stream_refs: FileRange, // location of the stream references array
+    pub object_refs: FileRange, // location of the object references array
+    pub stream: FileRange,      // location of the zstd-compressed stream within the file
+    pub named_refs: FileRange,  // location of the compressed named references
+    pub content_type: U64,      // user can put whatever magic identifier they want there
```

Comment:
I thought you were looking at removing the content type? I think we should do that given the new stream type prefixes.

Reply:
There's one thing that makes me want to keep content type: some future "deep fsck" mode where we not only ensure basic connectivity at the repository level, but make sure that the splitstreams have the expected content.  We need to know what kind of file we're dealing with in order to do that.